src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -0,0 +1,1143 @@
+        batch_description: Option<(BatchDescription, Capability<Timestamp>)>,
+    }
+
+    impl State {
+        fn new(
+            worker_id: usize,
+            persist_writer: WriteHandle<SourceData, (), Timestamp, Diff>,
+            metrics: SinkMetrics,
+        ) -> Self {
+            let worker_metrics = metrics.for_worker(worker_id);
+
+            Self {
+                worker_id,
+                persist_writer,
+                corrections: OkErr::new(
+                    Correction::new(metrics.clone(), worker_metrics.clone()),
+                    Correction::new(metrics, worker_metrics),
+                ),
+                desired_frontiers: OkErr::new_frontiers(),
+                persist_frontiers: OkErr::new_frontiers(),
+                batch_description: None,
+            }
+        }
+
+        fn trace<S: AsRef<str>>(&self, message: S) {
+            let message = message.as_ref();
+            trace!(
+                worker = %self.worker_id,
+                desired_frontier = ?self.desired_frontiers.frontier().elements(),
+                persist_frontier = ?self.persist_frontiers.frontier().elements(),
+                batch_description = ?self.batch_description.as_ref().map(|(d, _)| d),
+                message,
+            );
+        }
+
+        fn advance_desired_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.desired_frontiers.ok, frontier) {
+                self.trace("advanced `desired` ok frontier");
+            }
+        }
+
+        fn advance_desired_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.desired_frontiers.err, frontier) {
+                self.trace("advanced `desired` err frontier");
+            }
+        }
+
+        fn advance_persist_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.ok, frontier) {
+                self.apply_persist_frontier_advancement();
+                self.trace("advanced `persist` ok frontier");
+            }
+        }
+
+        fn advance_persist_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.err, frontier) {
+                self.apply_persist_frontier_advancement();
+                self.trace("advanced `persist` err frontier");
+            }
+        }
+
+        /// Apply the effects of a previous `persist` frontier advancement.
+        fn apply_persist_frontier_advancement(&mut self) {
+            let frontier = self.persist_frontiers.frontier();
+
+            // We will only emit times at or after the `persist` frontier, so now is a good time to
+            // advance the times of stashed updates.
+            self.corrections.ok.advance_since(frontier.clone());
+            self.corrections.err.advance_since(frontier.clone());
+
+            // If the `persist` frontier is beyond the `lower` of the current batch description, we
```

Comment:
This is very nitpicky, but I think a more correct formulation would be: "If the `lower` is not beyond the `persist` frontier of the ...". Of maybe "if `lower < persist frontier`, we ..."

Because beyond is always tricky, and for example `t` is beyond `[t]`.

Reply:
Good catch!